libs/blocks/gnav/gnav-breadcrumbs.js

```diff
@@ -0,0 +1,156 @@
+  const defaultItem = {
+    path,
+    title: defaultTitle,
+  };
+  const res = await fetch(path);
+  if (!res.ok) return defaultItem;
+  const html = await res.text();
+  if (!html) return defaultItem;
+  const doc = new DOMParser().parseFromString(html, 'text/html');
+  if (!doc) return defaultItem;
+  const hideInNav = getMetadata('breadcrumbs-hide-page', doc) === 'on';
+  const hideLastPage = getMetadata('breadcrumbs-hide-last', doc) === 'on';
+  return {
+    path,
+    title: getItemTitle(doc, defaultTitle),
+    hideInNav,
+    hideLastPage,
+  };
+}
+
+async function getItems(pagePath) {
+  const items = [];
+  let path = pagePath;
+  while (path !== null) {
+    // eslint-disable-next-line no-await-in-loop
+    const item = await getItem(path);
+    if (item) items.push(item);
+    path = getParent(path);
+  }
+  items.reverse();
+  return items;
+}
+
+async function getBreadcrumbsFromUrl(pagePath) {
+  const items = await getItems(pagePath);
+  if (!items || items.length === 0) return null;
+  const ul = createTag('ul');
+  const last = items.length - 1;
+  items.forEach((item, idx) => {
+    if (item.hideInNav) return;
+    const li = createTag('li');
+    if (idx === last) {
+      if (!item.hideLastPage) {
+        li.textContent = `${item.title}`;
+        li.setAttribute('aria-current', 'page');
+      }
+    } else {
+      const a = createTag('a', { href: item.path });
+      a.textContent = `${item.title}`;
+      li.append(a);
+    }
+    ul.append(li);
+  });
+  return createTag('nav', { class: 'breadcrumbs', 'aria-label': 'Breadcrumb' }, ul);
+}
+
+function getBreadcrumbsFromPageBlock(element) {
+  const parent = element.querySelector(':scope .breadcrumbs');
+  if (!parent) return null;
+  const ul = parent.querySelector(':scope ul');
+  if (!ul) return null;
+  ul.querySelector(':scope li:last-of-type')?.setAttribute('aria-current', 'page');
+  const breadcrumbs = createTag('nav', { class: 'breadcrumbs', 'aria-label': 'Breadcrumb' }, ul);
+  parent.remove();
+  return breadcrumbs;
+}
+
+async function getBreadcrumbsFile(path) {
+  const parent = getParent(path);
+  if (!parent) return null;
+  const prefix = (parent === '/') ? '' : parent;
+  const resp = await fetch(`${prefix}/breadcrumbs.plain.html`);
+  if (resp.ok) return resp.text();
+  return getBreadcrumbsFile(parent);
+}
+
+async function getBreadcrumbsFromFile() {
+  const defaultTitle = document.location.pathname.split('/').pop();
+  const title = getItemTitle(document, defaultTitle);
+  const html = await getBreadcrumbsFile(document.location.pathname);
+  if (!html) return null;
+  const parser = new DOMParser();
+  const doc = parser.parseFromString(html, 'text/html');
+  const breadcrumbs = getBreadcrumbsFromPageBlock(doc);
+  if (!breadcrumbs) return null;
+  // add the breadcrumbs item for the current page
+  const hideLastPage = getMetadata('breadcrumbs-hide-last') === 'on';
+  if (!hideLastPage) {
+    const ul = breadcrumbs.querySelector(':scope ul');
```

Comment:
Saw you use`:scope` above as well. Since you're not using the direct child combinator, I'm wondering what the use of it here is?

Reply:
I'm using `:scope` to restrict the selectors to the breadcrumbs element. Otherwise the selectors would be applied to the entire document. It leads to the same result but I assume it's a bit more efficient.

Reply:
Interesting, that's something I've never considered. After a little bit of googling, I wasn't able to find any evidence that there's a performance improvement. Seems unnecessary/inelegant to me, but since it's equivalent like you said, it's fine to keep.